filebeat/input/journald/pkg/journalctl/reader.go

```diff
@@ -0,0 +1,299 @@
+//
+// More details can be found in the PR introducing this feature and related
+// issues. PR: https://github.com/elastic/beats/pull/40061.
+type Reader struct {
+	cmd      *exec.Cmd
+	dataChan chan []byte
+	errChan  chan error
+	logger   *logp.Logger
+	stdout   io.ReadCloser
+	stderr   io.ReadCloser
+	canceler input.Canceler
+	wg       sync.WaitGroup
+}
+
+// handleSeekAndCursor adds the correct arguments for seek and cursor.
+// If there is a cursor, only the cursor is used, seek is ignored.
+// If there is no cursor, then seek is used
+func handleSeekAndCursor(args []string, mode SeekMode, since time.Duration, cursor string) []string {
+	if cursor != "" {
+		args = append(args, "--after-cursor", cursor)
+		return args
+	}
+
+	switch mode {
+	case SeekSince:
+		sinceArg := time.Now().Add(since).Format(time.RFC3339Nano)
+		args = append(args, "--since", sinceArg)
+	case SeekTail:
+		args = append(args, "--since", "now")
+	case SeekHead:
+		args = append(args, "--no-tail")
+	}
+
+	return args
+}
+
+// New instantiates and starts a reader for journald logs.
+//
+// The Reader starts a `journalctl` process with JSON ouput to read the journal
+// entries. Units and syslog identifiers are passed using the corresponding CLI
+// flags, matchers are passed directly to `journalctl` then transports are added
+// as matchers using `_TRANSPORTS` key.
+//
+// `mode` defines the 'seek mode'. It indicates whether the journal should be
+// read from the tail, head or starting from the cursor. If a cursor is passed,
+// then the seek mode is ignored.
+//
+// To start reading from a relative time, use mode: SeekSince and since should
+// be a time.Duration relative to the current time to start reading the
+// journald.
+//
+// File is the journal file to be read, for the system journal use the string
+// `LOCAL_SYSTEM_JOURNAL`.
+//
+// It's the caller's responsibility to call `Close` on the reader to stop
+// the `journalctl` process.
+//
+// If `canceler` is cancelled, the reading goroutine is stopped and subsequent
+// calls to `Next` will return an error.
+func New(
+	logger *logp.Logger,
+	canceler input.Canceler,
+	units []string,
+	syslogIdentifiers []string,
+	transports []string,
+	matchers journalfield.IncludeMatches,
+	mode SeekMode,
+	cursor string,
+	since time.Duration,
+	file string) (*Reader, error) {
+
+	args := []string{"--utc", "--output=json", "--follow"}
+	if file != "" && file != localSystemJournalID {
+		args = append(args, "--file", file)
+	}
+
+	args = handleSeekAndCursor(args, mode, since, cursor)
+
+	for _, u := range units {
+		args = append(args, "--unit", u)
+	}
+
+	for _, i := range syslogIdentifiers {
+		args = append(args, "--identifier", i)
+	}
+
+	for _, m := range matchers.Matches {
+		args = append(args, m.String())
+	}
+
+	for _, m := range transports {
+		args = append(args, fmt.Sprintf("_TRANSPORT=%s", m))
+	}
+
+	logger.Debugf("Journalctl command: journalctl %s", strings.Join(args, " "))
```

Comment:
This could be an info log, it won't be that frequent, and it is probably necessary to debug any bug that is going to be filed against this input.

Reply:
That makes sense, I didn't want to expose this details for the internals at info level. Also I cannot see it having any sensitive information we would not want at info level.